microdnf/plugins.cpp

```diff
@@ -0,0 +1,156 @@
+    explicit PluginLibrary(const std::string & library_path);
+
+    ~PluginLibrary();
+
+private:
+    using TGetApiVersionFunc = decltype(&microdnf_plugin_get_api_version);
+    using TGetNameFunc = decltype(&microdnf_plugin_get_name);
+    using TGetVersionFunc = decltype(&microdnf_plugin_get_version);
+    using TNewInstanceFunc = decltype(&microdnf_plugin_new_instance);
+    using TDeleteInstanceFunc = decltype(&microdnf_plugin_delete_instance);
+    TGetApiVersionFunc get_api_version{nullptr};
+    TGetNameFunc get_name{nullptr};
+    TGetVersionFunc get_version{nullptr};
+    TNewInstanceFunc new_instance{nullptr};
+    TDeleteInstanceFunc delete_instance{nullptr};
+    utils::Library library;
+};
+
+PluginLibrary::PluginLibrary(const std::string & library_path) : library(library_path) {
+    get_api_version = reinterpret_cast<TGetApiVersionFunc>(library.get_address("microdnf_plugin_get_api_version"));
+    get_name = reinterpret_cast<TGetNameFunc>(library.get_address("microdnf_plugin_get_name"));
+    get_version = reinterpret_cast<TGetVersionFunc>(library.get_address("microdnf_plugin_get_version"));
+    new_instance = reinterpret_cast<TNewInstanceFunc>(library.get_address("microdnf_plugin_new_instance"));
+    delete_instance = reinterpret_cast<TDeleteInstanceFunc>(library.get_address("microdnf_plugin_delete_instance"));
+    iplugin_instance = new_instance(PLUGIN_API_VERSION);
+}
+
+PluginLibrary::~PluginLibrary() {
+    finish();
+    delete_instance(iplugin_instance);
+    iplugin_instance = nullptr;
+}
+
+Plugins::~Plugins() {
+    finish();
+    for (auto plugin = plugins.rbegin(), stop = plugins.rend(); plugin != stop; ++plugin) {
+        (*plugin).reset();
+    }
+}
+
+void Plugins::register_plugin(std::unique_ptr<Plugin> && plugin) {
+    auto logger = context->base.get_logger();
+    auto * iplugin = plugin->get_iplugin();
+    plugins.emplace_back(std::move(plugin));
+    auto name = iplugin->get_name();
+    auto version = iplugin->get_version();
+    logger->debug(fmt::format(
+        "Added plugin name=\"{}\", version=\"{}.{}.{}\"", name, version.major, version.minor, version.micro));
+}
+
+void Plugins::load_plugin(const std::string & file_path) {
+    auto logger = context->base.get_logger();
+    logger->debug(fmt::format("Loading plugin file=\"{}\"", file_path));
+    auto plugin = std::make_unique<PluginLibrary>(file_path);
+    auto * iplugin = plugin->get_iplugin();
+    plugins.emplace_back(std::move(plugin));
+    auto name = iplugin->get_name();
+    auto version = iplugin->get_version();
+    logger->debug(fmt::format(
+        "Loaded plugin name=\"{}\", version=\"{}.{}.{}\"", name, version.major, version.minor, version.micro));
+}
+
+void Plugins::load_plugins(const std::string & dir_path) {
+    auto logger = context->base.get_logger();
+    if (dir_path.empty())
+        throw std::runtime_error(_("Plugins::loadPlugins() dirPath cannot be empty"));
```

Comment:
- use brackets for `if` body
- the symbol names in the error message are in camelCase and I don't think it's a good message for the user (something like _"Cannot load plugins, provided plugin directory path is empty"_ would be better)
- do we even want this check? I assume if the path doesn't exist we'll get a `filesystem_error` a couple lines below, which would kind of cover this case too
- if we do want the check, a custom exception class would also be good to have

Looking at the next commit, there's already a check that the path exists, so this should also never occur.

Reply:
OK. I will remove it.